File: mamda/c_cpp/src/cpp/MamdaAuctionFields.cpp
/* $Id$
 *
 * OpenMAMA: The open middleware agnostic messaging API
 * Copyright (C) 2011 NYSE Technologies, Inc.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301 USA
 */

#include <mamda/MamdaAuctionFields.h>
#include <mamda/MamdaCommonFields.h>
#include <mama/mamacpp.h>

namespace Wombat
{
    static uint16_t  theMaxFid     = 0;
    static bool      initialised   = false;

    const MamaFieldDescriptor*  MamdaAuctionFields::UNCROSS_PRICE       = NULL;
    const MamaFieldDescriptor*  MamdaAuctionFields::UNCROSS_VOLUME      = NULL;
    const MamaFieldDescriptor*  MamdaAuctionFields::UNCROSS_PRICE_IND   = NULL;
    const MamaFieldDescriptor*  MamdaAuctionFields::AUCTION_TIME        = NULL;


    void MamdaAuctionFields::setDictionary (const MamaDictionary&  dictionary)
    {
        // Already initialised
        if (initialised)
        {
            return;
        }

        // Set the Common Fields
        if(!MamdaCommonFields::isSet ())
        {
            MamdaCommonFields::setDictionary (dictionary);
        }

        UNCROSS_PRICE            = dictionary.getFieldByName ("wUncrossPrice");
        UNCROSS_VOLUME           = dictionary.getFieldByName ("wUncrossVolume");
        UNCROSS_PRICE_IND        = dictionary.getFieldByName ("wUncrossPriceInd");
        AUCTION_TIME             = dictionary.getFieldByName ("wAuctionTime");

        theMaxFid   = dictionary.getMaxFid();
        initialised = true;
    }

    void MamdaAuctionFields::reset ()
    {
        initialised         = false;

        //reset the Common Fields
        if (MamdaCommonFields::isSet ())
        {
            MamdaCommonFields::reset ();
        }

        theMaxFid           = 0;
        UNCROSS_PRICE       = NULL;
        UNCROSS_VOLUME      = NULL;
        UNCROSS_PRICE_IND   = NULL;
        AUCTION_TIME        = NULL;
    }

    bool MamdaAuctionFields::isSet ()
    {
        return initialised;
    }

    uint16_t  MamdaAuctionFields::getMaxFid ()
    {
        return theMaxFid;
    }

} // namespace

File: mamda/java/src/main/java/com/wombat/mamda/MamdaAuctionFields.java
/* $Id$
 *
 * OpenMAMA: The open middleware agnostic messaging API
 * Copyright (C) 2012 NYSE Technologies, Inc.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301 USA
 */

package com.wombat.mamda;

import com.wombat.mama.MamaFieldDescriptor;
import com.wombat.mama.MamaDictionary;

import java.util.Properties;

/**
* Maintains a cache of common update field descriptors.
*/
public class MamdaAuctionFields extends MamdaFields
{
    public static MamaFieldDescriptor  UNCROSS_PRICE        = null;
    public static MamaFieldDescriptor  UNCROSS_VOLUME       = null;
    public static MamaFieldDescriptor  UNCROSS_PRICE_IND    = null;

    public static int  MAX_FID = 0;    
    private static boolean initialised = false;

    /**
     * Set the dictionary for common update fields.
     * Maintains a cache of MamaFieldDescriptors for common update
     * related fields.
     * The <code>properties</code> parameter allows users
     * of the API to map the common dictionary names to
     * something else if they are being published under different
     * names.
     *
     * @param dictionary A reference to a valid MamaDictionary
     * @param properties A Properties object containing field mappings. (See
     * MamdaFields for more information)
     */
    public static void setDictionary (MamaDictionary  dictionary,
                                      Properties      properties)
    {
        // already initialised
        if (initialised)
        {
            return;
        }
        MamdaCommonFields.setDictionary (dictionary, null);

        UNCROSS_PRICE       = lookupField (dictionary, properties, "wUncrossPrice");
        UNCROSS_VOLUME      = lookupField (dictionary, properties, "wUncrossVolume");
        UNCROSS_PRICE_IND   = lookupField (dictionary, properties, "wUncrossPriceInd");     
        
        MAX_FID = dictionary.getMaxFid();        
        initialised = true;
    }

    private static MamaFieldDescriptor lookupField (MamaDictionary dictionary,
                                                    Properties     properties,
                                                    String         name)
    {
        return dictionary.getFieldByName (lookupFieldName (properties, name));
    }

    public static boolean isSet ()
    {
        return initialised;
    }
    
    public static void reset ()
    {
        if (MamdaCommonFields.isSet())
        {
            MamdaCommonFields.reset ();
        }
        
        MAX_FID             = 0;
        UNCROSS_PRICE       = null;
        UNCROSS_VOLUME      = null;
        UNCROSS_PRICE_IND   = null;

        initialised         = false;
    }

    public static int getMaxFid ()
    {
        return MAX_FID;
    }
}



